crates/cli/src/bin/wasm-bindgen-test-runner/headless.rs

```diff
@@ -0,0 +1,444 @@
+    // enough diagnostic info to see what went wrong (if anything).
+    let output = client.text(&id, &output)?;
+    let logs = client.text(&id, &logs)?;
+    let errors = client.text(&id, &errors)?;
+
+    if output.contains("test result: ") {
+        println!("{}", output);
+
+        // If the tests harness finished (either successfully or unsuccessfully)
+        // then in theory all the info needed to debug the failure is in its own
+        // output, so we shouldn't need the driver logs to get printed.
+        print_driver_stdio.set(false);
+    } else {
+        println!("failed to detect test as having been run");
+        if output.len() > 0 {
+            println!("output div contained:\n{}", tab(&output));
+        }
+    }
+    if logs.len() > 0 {
+        println!("console.log div contained:\n{}", tab(&logs));
+    }
+    if errors.len() > 0 {
+        println!("console.log div contained:\n{}", tab(&errors));
+    }
+
+    Ok(())
+}
+
+enum Driver {
+    Gecko(PathBuf),
+    Safari(PathBuf),
+    Chrome(PathBuf),
+}
+
+impl Driver {
+    fn find() -> Result<(Driver, Vec<String>), Error> {
+        let env_args = |name: &str| {
+            env::var(format!("{}_ARGS", name.to_uppercase()))
+                .unwrap_or_default()
+                .split_whitespace()
+                .map(|s| s.to_string())
+                .collect::<Vec<_>>()
+        };
+
+        let drivers = [
+            ("geckodriver", Driver::Gecko as fn(PathBuf) -> Driver),
+            ("safaridriver", Driver::Safari as fn(PathBuf) -> Driver),
+            ("chromedriver", Driver::Chrome as fn(PathBuf) -> Driver),
+        ];
+
+        // First up, if env vars like GECKODRIVER are present, use those to
+        // allow forcing usage of a particular driver.
+        for (driver, ctor) in drivers.iter() {
+            let env = driver.to_uppercase();
+            let path = match env::var_os(&env) {
+                Some(path) => path,
+                None => continue,
+            };
+            return Ok((ctor(path.into()), env_args(driver)))
+        }
+
+        // Next, check PATH. If we can find any supported driver, use that by
+        // default.
+        for path in env::split_paths(&env::var_os("PATH").unwrap_or_default()) {
+            let found = drivers
+                .iter()
+                .find(|(name, _)| {
+                    path.join(name)
+                        .with_extension(env::consts::EXE_EXTENSION)
+                        .exists()
+                });
+            let (name, ctor) = match found {
+                Some(p) => p,
+                None => continue,
+            };
+            return Ok((ctor(name.into()), env_args(name)))
+        }
+
+        // TODO: download an appropriate driver? How to know which one to
+        //       download?
+
+        bail!("failed to find a suitable WebDriver binary to drive headless \
+               testing")
```

Comment:
This message should include info about the env vars that we check and how we look for `foodriver` on `$PATH` or else we will drive users crazy